src/cryptographic_primitives/hashing/ext.rs

```diff
@@ -102,8 +102,22 @@ where
     }
 
     fn result_scalar<E: Curve>(self) -> Scalar<E> {
-        let n = self.result_bigint();
-        Scalar::from_bigint(&n)
+        let scalar_len = E::Scalar::SCALAR_LENGTH;
+        assert!(
+            Self::output_size() >= scalar_len,
+            "Output size of the hash({}) is smaller than the scalar length({})",
+            Self::output_size(),
+            scalar_len
+        );
+        // Try and increment.
+        for i in 0u32.. {
+            let starting_state = self.clone();
```

Comment:
I noticed that we have to add `+ Clone` in a lot of places because of this `.clone()`. It's not really critical, but we can eliminate this by finding first `H(data || 0x1 || 0x2 || ... || i)` instead of `H(data || i)`. Are there any downsides of this approach?

Reply:
good question, I'll try to think if this can introduce any attacks.
(small note, AFAIK all hash functions implement Clone)

Reply:
Yeah, all hashes must implement Clone. My suggestion is only about saving few keystrokes. If you think it might introduce any attacks, let's leave it as is

Reply:
Btw do we even need this constraint? According to the documentation, Digest trait is implemented for any D that `D: Clone + FixedOutput + ...`. Can't Rust deduce that any Digest implements Clone? See https://docs.rs/digest/0.9.0/digest/trait.Digest.html#impl-Digest

Reply:
It's implemented for it doesn't inherit it :/ 
I can instead do `D: FixedOutput + Update + Clone` but sadly I can't remove clone